File: Binsearch.java

class Binsearch {

	public int binsearch(int[] array, int el) {
		return binsearch(array, 0, array.length-1, el);
	}

	public int calculateHalf(int start, int end) {
		return start + (end-start)/2;
		// wow, unsigned bitshift! :)
		//return (start+end) >>>1;
	}

	private int binsearch(int[] array, int start, int end, int el) {
		if ( end < start ) return -1;

		//int half = start + (end-start)/2;
		int half = calculateHalf(start, end);

		if ( array[half] == el )
			return half;
		if ( start == end )
			return -1;

		if ( array[half] > el )
			return binsearch(array, 0, half-1, el);
		else
			return binsearch(array, half+1, end, el);
	}

	Binsearch() {
	}

	Binsearch(String[] args) {
		if ( args.length == 0 ) 
			System.exit(-1);

		int[] array = new int[args.length];

		System.out.println("argc (and el to look for): " + args.length);

		for ( int i = 0 ; i < args.length ; i++ )
			array[i] = Integer.parseInt(args[i]);

		System.out.println("array (must be sorted): ");
		for ( int i = 0 ; i < args.length ; i++ )
			System.out.print(array[i] + " ");
		System.out.println();

		int ret = binsearch(array, args.length);

		if ( ret == -1 )
			System.out.println("Not Found");
		else
			System.out.println("Found at " + ret);
	}

	public static void main (String[] args) {
		new Binsearch(args);
	}

}

File: binsearch.c
#include <stdio.h>
#include <stdlib.h>

int binsearch(int array[], size_t start, size_t end, int el) {
	int half = start + (end-start)/2;

	if ( array[half] == el )
		return half;
	if ( start == end )
		return -1;

	if ( array[half] > el )
		return binsearch(array, 0, half-1, el);
	else
		return binsearch(array, half+1, end, el);
}

int main ( int argc, char **argv ) { 
	if ( argc == 1 ) 
		return -1;
	int i;
	int el = argc;
	int array[argc-1];
	printf("argc and number to look for: %d\n", argc);
	for ( i = 0 ; i < argc-1 ; i++ )
		array[i] = atoi(argv[i+1]);

	printf("array (must be sorted): ");
	for ( i = 0 ; i < argc-1 ; i++ )
		printf("%d ", array[i]);
	printf("\n");

	printf("end: %d\n", argc-2);

	/* here core */
	int ret = binsearch(array, 0, argc-2, el);
	if ( ret == -1 )
		printf("not found\n");
	else
		printf("found at %d\n", ret);

	return 0;
}
